### PR TITLE
Sqlite Attach Database

### PR DESCRIPTION
**Pull request by elmpp**
### Description of change

Adds support for SQLite [attach databases](https://sqlite.org/lang_attach.html). This allows database files additional to the default "main" file that TypeORM currently supports. 

This implementation is quite similar to current support for multi-database (postgres, mysql) but without the concept of schema paths. The databases are configured in the same way also - as entity decorators; albeit using filepath strings

The bulk of the complexity comes from platform filepath differences - i felt a big requirement would be that migration files, synchronised file locations etc would be portable between windows/posix. As a result there are a few path utils (with tests) to achieve that. 

Notes:
 - ~Foreign keys aren't supported cross-database so are turned off if any dbs are attached. It could be possible to still support them within the same db. What do you think?~ Foreign keys between tables of same db are included
 - ~i wasn't able to build 'better-sqlite' on my machine so hoping CI will test that~
 - adds `ATTACH` statements to migrations

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run lint` passes with this change
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [x] There are new or updated unit tests validating the change
- [x] Documentation has been updated to reflect this change
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: src/metadata/EntityMetadata.ts
Comment (nebkat) [inline src/metadata/EntityMetadata.ts:866]: At this point lets just abstract `buildTablePath()` for all drivers and pass `tableName`, `schema` and `database`. The function could be made optional also and default to using `tableName` by default.

File: src/query-builder/QueryBuilder.ts
Comment (nebkat) [inline src/query-builder/QueryBuilder.ts:88]:
```suggestion
```

File: src/driver/sqlite/SqliteDriver.ts
Comment (nebkat) [inline src/driver/sqlite/SqliteDriver.ts:95]: Missing semicolons throughout this function (and a few others).

File: src/driver/sqlite-abstract/AbstractSqliteQueryRunner.ts
Comment (nebkat) [inline src/driver/sqlite-abstract/AbstractSqliteQueryRunner.ts:1171]: Can `_` be used for unused vars?

File: src/driver/sqlite-abstract/AbstractSqliteQueryRunner.ts
Comment (nebkat) [inline src/driver/sqlite-abstract/AbstractSqliteQueryRunner.ts:1225]: Should be careful here, `split('.')` could return more than 2. Better to do something like
```suggestion
        let dotIndex = tablePath.indexOf(".");
        return (dotIndex >= 0) ? [tablePath.substr(0, dotIndex), tablePath.substr(dotIndex + 1)] : [undefined, tablePath];
```

Comment (nebkat) [reply src/driver/sqlite-abstract/AbstractSqliteQueryRunner.ts:1225]: Also make sure all strings are in double quotes rather than single, should be picked up by ESLint in IDE?

**Review by nebkat — CHANGES_REQUESTED** [review]

File: src/metadata/EntityMetadata.ts
Comment (elmpp) [reply src/metadata/EntityMetadata.ts:866]: I didn't want to do this outright in my first PR because i'm still unclear the differences between `tableName` and `tablePath`.

I think it needs a bit more thought (and probably needs more consistent escaping strategies in certain places)

File: src/driver/sqlite-abstract/AbstractSqliteQueryRunner.ts
Comment (elmpp) [inline src/driver/sqlite-abstract/AbstractSqliteQueryRunner.ts:1177]: would be nice if this was allowable?

File: src/driver/sqlite-abstract/AbstractSqliteQueryRunner.ts
Comment (nebkat) [reply src/driver/sqlite-abstract/AbstractSqliteQueryRunner.ts]: I checked now the syntax is actually to not put in any variable:

`[ , tableNameNew]`

File: src/metadata/EntityMetadata.ts
Comment (nebkat) [reply src/metadata/EntityMetadata.ts]: The only difference should be that `tablePath` contains the full database/schema name, while `tableName` does not. I think drivers have also so far been responsible for escaping, but is there something else I'm missing here?

Comment (elmpp) [reply src/metadata/EntityMetadata.ts]: the concept of `tablePath` vs `tableName` seems a little blurry throughout - it differs from here in the `EntityMetadata`, through the `Table` and `TableIndex` instances. Also, the same driver instance can expect differing values based on the platform (the `AbstractSqliteDriver/QueryRunner`). e.g. `sql.js` does not support tablePaths on indexes etc

Comment (elmpp) [reply src/metadata/EntityMetadata.ts]: tl;dr; i'd rather not include cross-driver changes in this PR

File: src/connection/BaseConnectionOptions.ts
Comment (imnotjames) [inline src/connection/BaseConnectionOptions.ts:116]: What happens when there's no ormconfig?

Why does this feature operate differently from how the standard database path resolution works for sqlite?

File: src/connection/ConnectionOptionsReader.ts
Comment (imnotjames) [inline src/connection/ConnectionOptionsReader.ts:182]: The attaching should probable be defined within the connection options for sqlite instead of in random entities so they are in the same place as the main file's database.  The entities could then reference the alias instead of the filename.

File: src/driver/sqlite-abstract/AbstractSqliteDriver.ts
Comment (imnotjames) [inline src/driver/sqlite-abstract/AbstractSqliteDriver.ts:50]: can this be protected?

File: src/driver/sqlite/SqliteDriver.ts
Comment (imnotjames) [inline src/driver/sqlite/SqliteDriver.ts:174]: Probably can drop this function and instead use `mkdirp` directly

**Review by imnotjames — CHANGES_REQUESTED** [review]
We aren't using yarn for this project, please remove references to it.

**Comment by pleerock** [comment]
I'll close it since there is no answer from author. Feel free to re-open.

**Comment by elmpp** [comment]
@pleerock  Hi there - I must have overlooked the second review of @imnotjames . If I find some time this week to respond is there a decent chance it could be considered? (there been a bit of a wait)

Not sure how to reopen this PR tbh either ?

**Comment by imnotjames** [comment]
You should be able to reopen it yourself.

**Comment by elmpp** [comment]
I see no option to reopen - only the ability to "reference in a new issue". I'd like to keep the comments etc if possible. Weirdly, i'm also unable to upload images in this reply

**Comment by elmpp** [comment]
I have rebased the branch to latest master and added more tests. 

I thought adding commits would give me the option to reopen but looks like not 

I'm going to open another PR with a reference to here. I assume you are filtering closed PR comments 👍

**Pull request closed without merging**
